cpp/tests/test_communicator.cpp

```diff
@@ -0,0 +1,81 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <rmm/mr/device/cuda_memory_resource.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <rapidsmpf/buffer/buffer.hpp>
+#include <rapidsmpf/buffer/resource.hpp>
+#include <rapidsmpf/communicator/communicator.hpp>
+
+#include "environment.hpp"
+#include "utils.hpp"
+
+class BaseCommunicatorTest : public ::testing::Test {
+  protected:
+    void SetUp() override {
+        comm = GlobalEnvironment->comm_.get();
+        mr = std::unique_ptr<rmm::mr::device_memory_resource>(
+            new rmm::mr::cuda_memory_resource{}
+        );
+        br = std::make_unique<rapidsmpf::BufferResource>(mr.get());
+        stream = rmm::cuda_stream_default;
+        GlobalEnvironment->barrier();
+    }
+
+    void TearDown() override {
+        GlobalEnvironment->barrier();
+    }
+
+    virtual rapidsmpf::MemoryType memory_type() = 0;
+
+    rapidsmpf::Communicator* comm;
+    std::unique_ptr<rmm::mr::device_memory_resource> mr;
+    rmm::cuda_stream_view stream;
+    std::unique_ptr<rapidsmpf::BufferResource> br;
+};
+
+class BasicCommunicatorTest
+    : public BaseCommunicatorTest,
+      public ::testing::WithParamInterface<rapidsmpf::MemoryType> {
+  protected:
+    rapidsmpf::MemoryType memory_type() override {
+        return GetParam();
+    }
+};
+
+INSTANTIATE_TEST_CASE_P(
+    BasicCommunicatorTest,
+    BasicCommunicatorTest,
+    testing::Values(rapidsmpf::MemoryType::HOST, rapidsmpf::MemoryType::DEVICE),
+    [](const testing::TestParamInfo<rapidsmpf::MemoryType>& info) {
+        return "memory_type_" + std::to_string(static_cast<int>(info.param));
+    }
+);
+
+TEST_P(BasicCommunicatorTest, SendToSelf) {
+    if (GlobalEnvironment->type() != TestEnvironmentType::MPI) {
+        GTEST_SKIP() << "Unsupported send to self";
+    }
```

Comment:
Ah! I was wondering how this PR passed CI. This test fails in ucx :slightly_smiling_face: